core/src/test/java/org/apache/calcite/sql2rel/RelFieldTrimmerTest.java

```diff
@@ -306,4 +314,139 @@
+    final RelNode root =
+        builder.scan("EMP")
+            .project(builder.field("EMPNO"), builder.field("ENAME"), builder.field("DEPTNO"))
+            .exchange(RelDistributions.SINGLETON)
+            .filter(
+                builder.call(SqlStdOperatorTable.GREATER_THAN,
+                    builder.field("EMPNO"), builder.literal(100)))
+            .project(builder.field("EMPNO"), builder.field("ENAME"))
+            .build();
+
+    final HepProgram hepProgram = new HepProgramBuilder()
+        .addRuleInstance(ProjectToCalcRule.INSTANCE)
+        .addRuleInstance(FilterToCalcRule.INSTANCE)
+        .addRuleInstance(CalcMergeRule.INSTANCE).build();
+
+    final HepPlanner hepPlanner = new HepPlanner(hepProgram);
+    hepPlanner.setRoot(root);
+    final RelNode relNode = hepPlanner.findBestExp();
+    final RelFieldTrimmer fieldTrimmer = new RelFieldTrimmer(null, builder);
+    final RelNode trimmed = fieldTrimmer.trim(relNode);
+
+    final String expected = ""
+        + "LogicalCalc(expr#0..1=[{inputs}], expr#2=[100], expr#3=[>($t0, $t2)], proj#0."
+        + ".1=[{exprs}], $condition=[$t3])\n"
+        + "  LogicalExchange(distribution=[single])\n"
+        + "    LogicalCalc(expr#0..1=[{inputs}], proj#0..1=[{exprs}])\n"
+        + "      LogicalProject(EMPNO=[$0], ENAME=[$1])\n"
+        + "        LogicalTableScan(table=[[scott, EMP]])\n";
+    assertThat(trimmed, hasTree(expected));
+  }
+
+  @Test void testCalcWithHints() {
+    final RelHint calcHint = RelHint.builder("resource").build();
+    final RelBuilder builder = RelBuilder.create(config().build());
+    builder.getCluster().setHintStrategies(
+        HintStrategyTable.builder().hintStrategy("resource", HintPredicates.CALC).build());
+    final RelNode original =
+        builder.scan("EMP")
+            .project(
+                builder.field("EMPNO"),
+                builder.field("ENAME"),
+                builder.field("DEPTNO")
+            ).hints(calcHint)
+            .sort(builder.field("EMPNO"))
+            .project(builder.field("EMPNO"))
+            .build();
+
+    final HepProgram hepProgram = new HepProgramBuilder()
+        .addRuleInstance(ProjectToCalcRule.INSTANCE)
+        .build();
+    final HepPlanner hepPlanner = new HepPlanner(hepProgram);
+    hepPlanner.setRoot(original);
+    final RelNode relNode = hepPlanner.findBestExp();
+
+    final RelFieldTrimmer fieldTrimmer = new RelFieldTrimmer(null, builder);
+    final RelNode trimmed = fieldTrimmer.trim(relNode);
+
+    final String expected = ""
+        + "LogicalCalc(expr#0=[{inputs}], EMPNO=[$t0])\n"
+        + "  LogicalSort(sort0=[$0], dir0=[ASC])\n"
+        + "    LogicalCalc(expr#0=[{inputs}], EMPNO=[$t0])\n"
+        + "      LogicalProject(EMPNO=[$0])\n"
+        + "        LogicalTableScan(table=[[scott, EMP]])\n";
+    assertThat(trimmed, hasTree(expected));
+
+    assertTrue(original.getInput(0).getInput(0) instanceof Project);
```

Comment:
minor: I think in this test it would also make sense to verify that the original Calc (`relNode.getInput(0).getInput(0)`) contains the hints.

Reply:
Thanks, Add the verification logic of the original Calc contains the hints.